### PR TITLE
add help page listing available macros

### PR DESCRIPTION
**Pull request by abizer**
looks like this:

macros

### CODE REVIEW

**Comment by dkess** [comment]
Are we enforcing that these are image links? If not, we need a way to handle non-image macros.

**Comment by abizer** [comment]
there's no good way to enforce that they are image links, I don't think. Whitelisting file extensions is the easiest to implement but pretty fragile (e.g. #m nattofriends -> https://yelp.fuqu.jp, which just returns an image), downloading the image and trying to do some imagemagick shit to see if it is valid seems extremely bad. Do you have any suggestions? Right now, if the links are not images, it just shows the broken image link, which doesn't actually break anything.

**Comment by dkess** [comment]
Put the text in the `alt` attribute. If the image fails to load, that will show up in place of the image. See the source of https://www.ocf.berkeley.edu/~dkessler/image_error

**Comment by abizer** [comment]
I already am - see macros.html#L40

File: ircbot/plugin/help.py [hunk 1]
```diff
@@ -24,24 +30,32 @@ def build_request_handler(bot):
...
-                rendered = jinja_env.get_template('plugin/templates/help.html').render(
-                    plugins=sorted(plugins.items(), key=lambda p: p[0].__name__),
-                ).encode('utf8')
-
-                self.send_response(200, 'Okay')
-                self.send_header('Content-Length', len(rendered))
-                self.send_header('Content-Type', 'text/html; charset=utf-8')
+                self.render_response(
+                    'plugin/templates/help.html',
+                    plugins=sorted(plugins.items(), key=lambda p: p[0].__name__)
```
Comment (jvperrin) [inline ircbot/plugin/help.py]: nit: add a trailing comma

File: ircbot/plugin/help.py [hunk 2]
```diff
@@ -24,24 +30,32 @@ def build_request_handler(bot):
...
+                )
+            elif self.path == '/macros':
+                self.render_response(
+                    'plugin/templates/macros.html',
+                    macros=bot.plugins['macros'].list(bot)
```
Comment (jvperrin) [inline ircbot/plugin/help.py]: nit: add a trailing comma here too

File: ircbot/plugin/macros.py [hunk 3]
```diff
@@ -71,3 +71,12 @@ def delete(bot, msg):
             (slug,)
         )
         msg.respond('macro `{}` has been deleted.'.format(slug))
+
+
+def list(bot):
+    """list all macros for help page"""
```
Comment (chriskuehl) [inline ircbot/plugin/macros.py]: this should be a real sentence (capital first letter, end with a period) per https://www.python.org/dev/peps/pep-0257/#one-line-docstrings

File: ircbot/plugin/macros.py [hunk 4]
```diff
@@ -71,3 +71,12 @@ def delete(bot, msg):
...
+
+    with db.cursor(password=bot.mysql_password) as c:
+        c.execute('SELECT slug, link FROM macros ORDER BY slug DESC')
```
[the remark this answers is not in the document]
Comment (abizer) [reply ircbot/plugin/macros.py]: yeah good point I forgot to take this out from testing

File: ircbot/plugin/macros.py [hunk 5]
```diff
@@ -71,3 +71,12 @@ def delete(bot, msg):
...
+        for entry in c.fetchall():
+            yield entry['slug'], entry['link']
```
[the remark this answers is not in the document]
Comment (abizer) [reply ircbot/plugin/macros.py]: interesting, thanks for noting this.

**Pull request merged**